docs/proposals/resource-metrics-api.md

```diff
@@ -0,0 +1,166 @@
+
+HPA uses the latest value of cpu usage as an average aggregated across 1 minute
+(the window may change in the future). The data for a given set of pods
+(defined either by pod list or label selector) should be accesible in one request
+due to performance issues.
+
+#### Scheduler
+
+Scheduler in order to schedule best-effort pods requires node level resource usage metrics
+as an average aggreated across 1 minute (the window may change in the future).
+The metrics should be available for all resources supported in the scheduler.
+
+## Other considered use cases
+
+This paragraph describes the other considered use cases and explains why they are out
+of the scope MVP version.
+
+#### Custom metrics in HPA
+
+HPA requires the latest value of application level metrics.
+
+The effort is suspended and it's no clear whether application level metrics should be
+available in API server so the use case initially won't be supported.
+
+#### Ubernetes
+
+In order to schedule best-effort pods in Ubernetes cluster the Ubernetes scheduler
+requires the same metrics that the Kubernetes scheduler. To effort is in planning phase
+so it's out the scope of Resource Metrics API effort however the support should be sufficient
+once Kubernetes scheduler is supported.
+
+#### kubectl top
+
+This feature is not yet specified/implemented although it seems reasonable to support the following options:
+- ```kubectl top node <id>``` - list all pods running on the given node accompanied with theirs cpu/mem usage
+- ```kubectl top pod <id>``` - print cpu usage stats for the given pod
+- ```kubectl top pod --selector=<label-query>``` - list all pods defined by the label query accompanied with theirs cpu/mem usage
+
+Since there are not hard requirements in this case it will be not supported initally in the API although
+it will be probably possible to provide a reasonable implementation of the feature anyway.
+
+#### Kubernetes dashboard
+
+[Kubernetes dashboard](https://github.com/kubernetes/dashboard) in order to draw graps requires resource usage
+in timeseries format from relatively long period of time. The aggreations should be also possible on various levels
+including replication controllers, deployments, services, etc.
+
+Since the use case is complicated it will not be supported initally in the API and they will query Heapster
+directly using some custom API there.
+
+## Proposed API
+
+Initially the metrics API will be in a separate [API group](api-group.md) called ```metrics```.
+Later if we decided to have Node and Pod in different API groups also
+NodeMetrics and PodMetrics should be in different API groups.
+
+#### Schema
+
+The proposed schema is as follow. Each top-level object has `TypeMeta` and `ObjectMeta` fields
+to be compatible with Kubernetes API standards.
+
+```go
+type NodeMetrics struct {
+  unversioned.TypeMeta
+  ObjectMeta
+
+  Timestamp unversioned.Time
+  Window    unversioned.Duration
+  
+  Usage map[ResourceType]Resource.Quantity
+}
+
```

Comment:
in a separate email, @piosz asked:
> How should we account CPU usage on node level?
>
> AFAIU the code, currently in stats summary API we export just the cpu usage of the whole node. > On the other hand from scheduler perspective it might make sense to use the sum of cpu usage of all pods running on the machine especially assuming we will start supporting Allocatable feature.
>
> Technically as for now there is no difference. There is aggregating logic already implemented in Heapster so adding yet another metric to aggregate is a 2-lines change.

If the scheduler is just using a "least loaded" heuristic for spreading work, then whole node cpu usage is sufficient.   It is just a rough heuristic, so we should do the simplest thing, I think.

Regarding unaccounted cpu usage: If some node has a bunch of CPU usage that is not accounted to any container, (e.g. kernel threads or processes in root cpu cgroup), we still want to avoid that node.